File: SuperiorPackGroup/CarriersXtraForm.vb
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Public Class CarriersXtraForm

    Private m_Carriers As CarriersBLL
    Private m_CurrentCarrierID As Nullable(Of Integer)

    Private Sub CarriersXtraForm_FormClosing(ByVal sender As Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles Me.FormClosing

        If Me.carriersSearchGridControl.Enabled = False Then
            Select Case MessageBox.Show("Do you want to save changes?", "Save Changes", MessageBoxButtons.YesNoCancel, MessageBoxIcon.Question)
                Case Windows.Forms.DialogResult.Yes
                    If SaveChanges() = False Then
                        e.Cancel = True
                    End If
                Case Windows.Forms.DialogResult.No
                    CancelChanges()
                Case Windows.Forms.DialogResult.Cancel
                    e.Cancel = True
            End Select
        End If

    End Sub

    Private Sub CarriersXtraForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        Me.Cursor = Cursors.WaitCursor
        m_Carriers = New CarriersBLL

        BindCarrierSearch()

        CheckPermissions()

        Utilities.MakeFormReadOnly(Me.carriersXtraTabPage, True)
        Me.carriersSearchGridControl.Enabled = True

        cancelBarButtonItem.Enabled = False
        saveBarButtonItem.Enabled = False

        Me.Cursor = Cursors.Default

    End Sub

    Public Sub BindCarrierSearch()

        'Me.carrierListBoxControl.DisplayMember = "CarrierName"
        'Me.carrierListBoxControl.ValueMember = "CarrierID"
        Me.carriersSearchGridControl.DataSource = m_Carriers.GetCarrierIDAndNames()

    End Sub

    Private Function SaveChanges() As Boolean

        'Required to get back to current carrier when entering a new carrier.
        Dim selectedID As Integer = -1

        If m_CurrentCarrierID.HasValue = False Then
            m_CurrentCarrierID = m_Carriers.GetNewCarrierID()
        Else
            selectedID = carriersSearchGridView.FocusedRowHandle
        End If

        Try
            Dim firstName, lastName As String
            If Me.firstNameTextEdit.EditValue Is Nothing OrElse Convert.ToString(Me.firstNameTextEdit.EditValue) = Me.firstNameTextEdit.Properties.NullText Then
                firstName = String.Empty
            Else
                firstName = Convert.ToString(Me.firstNameTextEdit.EditValue)
            End If
            If Me.lastNameTextEdit.EditValue Is Nothing OrElse Convert.ToString(Me.lastNameTextEdit.EditValue) = Me.lastNameTextEdit.Properties.NullText Then
                lastName = String.Empty
            Else
                lastName = Convert.ToString(Me.lastNameTextEdit.EditValue)
            End If
            If m_Carriers.UpdateCarrier(m_CurrentCarrierID.Value, Me.nameTextEdit.Text, firstName, lastName, Me.addressTextEdit.Text, Me.cityTextEdit.Text, Me.stateTextEdit.Text, _
                    Me.postalTextEdit.Text, Me.phoneTextEdit.Text, Me.emailTextEdit.Text, Me.faxTextEdit.Text, Me.noteMemoEdit.Text) <> True Then
                MessageBox.Show("The carrier was not updated succesfully.", "Error Encountered", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Return False
            End If
        Catch ex As ApplicationException
            MessageBox.Show(ex.Message, "Error Encountered", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return False
        End Try

        BindCarrierSearch()
        Me.carriersSearchGridView.FocusedRowHandle = selectedID
        Return True

    End Function

    Private Sub saveSimpleButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs)

        SaveChanges()

    End Sub

    Private Sub carrierListBoxControl_SelectedValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs)

        'If Me.carrierListBoxControl.SelectedIndex <> -1 Then
        '    BindCarrierControls(CType(Me.carrierListBoxControl.SelectedValue, Integer))
        'End If

    End Sub

    Public Sub BindCarrierControls(ByVal carrierID As Integer)

        Dim carrier As SPG.CarriersRow

        carrier = CType(m_Carriers.GetCarrierByID(carrierID).Rows(0), SPG.CarriersRow)
        m_CurrentCarrierID = carrier.CarrierID
        Me.nameTextEdit.Text = carrier.CarrierName
        Me.firstNameTextEdit.Text = carrier.ContactFirstName
        Me.lastNameTextEdit.Text = carrier.ContactLastName
        Me.addressTextEdit.Text = carrier.Address
        Me.cityTextEdit.Text = carrier.City
        Me.stateTextEdit.Text = carrier.State
        Me.postalTextEdit.Text = carrier.Postal
        Me.phoneTextEdit.Text = carrier.Phone
        Me.emailTextEdit.Text = carrier.Email
        Me.faxTextEdit.Text = carrier.Fax
        Me.noteMemoEdit.Text = carrier.Note

    End Sub

    Private Sub addSimpleButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs)

        m_CurrentCarrierID = Nothing
        Me.nameTextEdit.Text = String.Empty
        Me.firstNameTextEdit.Text = Nothing
        Me.lastNameTextEdit.Text = Nothing
        Me.addressTextEdit.Text = String.Empty
        Me.cityTextEdit.Text = String.Empty
        Me.stateTextEdit.Text = String.Empty
        Me.postalTextEdit.Text = String.Empty
        Me.phoneTextEdit.Text = String.Empty
        Me.emailTextEdit.Text = String.Empty
        Me.faxTextEdit.Text = String.Empty
        Me.noteMemoEdit.Text = String.Empty

    End Sub

    Private Sub cancelSimpleButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs)

        'If Me.carrierListBoxControl.SelectedIndex <> -1 Then
        '    BindCarrierControls(CType(Me.carrierListBoxControl.SelectedValue, Integer))
        'End If

    End Sub

    Private Sub carriersSearchGridView_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles carriersSearchGridView.Click

        If (Me.carriersSearchGridView.CalcHitInfo(Me.carriersSearchGridControl.PointToClient(Control.MousePosition)).HitTest = GridHitTest.RowCell) Then
            Me.BindCarrierControls(CType(Me.carriersSearchGridView.GetFocusedRowCellValue(Me.carrierIDGridColumn), Integer))
        End If

    End Sub

    Private Sub addBarButtonItem_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles addBarButtonItem.ItemClick

        m_CurrentCarrierID = Nothing
        Me.nameTextEdit.Text = String.Empty
        Me.firstNameTextEdit.Text = Nothing
        Me.lastNameTextEdit.Text = Nothing
        Me.addressTextEdit.Text = String.Empty
        Me.cityTextEdit.Text = String.Empty
        Me.stateTextEdit.Text = String.Empty
        Me.postalTextEdit.Text = String.Empty
        Me.phoneTextEdit.Text = String.Empty
        Me.emailTextEdit.Text = String.Empty
        Me.faxTextEdit.Text = String.Empty
        Me.noteMemoEdit.Text = String.Empty

        Me.carriersSearchGridView.FocusedRowHandle = -1
        Utilities.MakeFormReadOnly(Me.carriersXtraTabPage, False)
        Me.carriersSearchGridControl.Enabled = False
        CheckPermissions()

        Me.cancelBarButtonItem.Enabled = True
        Me.saveBarButtonItem.Enabled = True
        Me.editBarButtonItem.Enabled = False
        Me.addBarButtonItem.Enabled = False

    End Sub

    Private Sub editBarButtonItem_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles editBarButtonItem.ItemClick

        If m_CurrentCarrierID Is Nothing Then
            MessageBox.Show("Please select a carrier to edit.")
            Exit Sub
        End If

        Utilities.MakeFormReadOnly(Me.carriersXtraTabPage, False)
        Me.carriersSearchGridControl.Enabled = False
        CheckPermissions()

        Me.cancelBarButtonItem.Enabled = True
        Me.saveBarButtonItem.Enabled = True
        Me.editBarButtonItem.Enabled = False
        Me.addBarButtonItem.Enabled = False

    End Sub

    Private Sub cancelBarButtonItem_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles cancelBarButtonItem.ItemClick

        CancelChanges()

        CheckPermissions()

        Utilities.MakeFormReadOnly(Me.carriersXtraTabPage, True)
        Me.carriersSearchGridControl.Enabled = True

        cancelBarButtonItem.Enabled = False
        saveBarButtonItem.Enabled = False

    End Sub

    Private Sub saveBarButtonItem_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles saveBarButtonItem.ItemClick

        If SaveChanges() Then

            CheckPermissions()

            Utilities.MakeFormReadOnly(Me.carriersXtraTabPage, True)
            Me.carriersSearchGridControl.Enabled = True

            cancelBarButtonItem.Enabled = False
            saveBarButtonItem.Enabled = False
        End If

    End Sub

    Public Function CheckPermissions() As Boolean

        Try
            Dim permission As SPG.UserPermissionsRow = New UserPermissionsBLL().GetUserPermissionsByUserName(My.Settings.UserName, "Carrierss Form")(0)

            Select Case permission.PermissionLevel
                Case "FULL", "Full"
                    Me.addBarButtonItem.Enabled = True
                    Me.editBarButtonItem.Enabled = True
                    'Me.deleteBarButtonItem.Enabled = True
                Case "ADD", "Add"
                    Me.addBarButtonItem.Enabled = True
                    Me.editBarButtonItem.Enabled = True
                Case "NONE", "None"
                    MessageBox.Show("You don't have permission to view this form.", "User Permissions", MessageBoxButtons.OK, MessageBoxIcon.Hand)
                    Me.Close()
                    Return False
                Case "VIEW", "View"
                    Me.addBarButtonItem.Enabled = False
                    Me.editBarButtonItem.Enabled = False
                    'Me.deleteBarButtonItem.Enabled = False
                Case "EDIT", "Edit"
                    Me.addBarButtonItem.Enabled = False
                    Me.editBarButtonItem.Enabled = True
                    'Me.deleteBarButtonItem.Enabled = False
            End Select

            Return True
        Catch ex As InvalidOperationException
            Return False
        End Try

    End Function

    Private Sub CancelChanges()

        If Me.m_CurrentCarrierID IsNot Nothing Then
            BindCarrierControls(m_CurrentCarrierID.Value)
        End If

    End Sub

End Class

File: SuperiorPackGroup/CarriersXtraForm.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;

namespace SuperiorPackGroup
{
	public partial class CarriersXtraForm
	{

		public CarriersXtraForm()
		{
			InitializeComponent();
		}

		private CarriersBLL m_Carriers;
		private int? m_CurrentCarrierID;

		private void CarriersXtraForm_FormClosing(object sender, FormClosingEventArgs e)
		{

			if (carriersSearchGridControl.Enabled == false)
			{
				switch (MessageBox.Show("Do you want to save changes?", "Save Changes", MessageBoxButtons.YesNoCancel, MessageBoxIcon.Question))
				{
					case DialogResult.Yes:
						if (SaveChanges() == false)
						{
							e.Cancel = true;
						}
						break;
					case DialogResult.No:
						CancelChanges();
						break;
					case DialogResult.Cancel:
						e.Cancel = true;
						break;
				}
			}

		}

		private void CarriersXtraForm_Load(System.Object sender, EventArgs e)
		{

			Cursor = Cursors.WaitCursor;
			m_Carriers = new CarriersBLL();

			BindCarrierSearch();

			CheckPermissions();

			Utilities.MakeFormReadOnly(carriersXtraTabPage, true);
			carriersSearchGridControl.Enabled = true;

			cancelBarButtonItem.Enabled = false;
			saveBarButtonItem.Enabled = false;

			Cursor = Cursors.Default;

		}

		public void BindCarrierSearch()
		{

			//Me.carrierListBoxControl.DisplayMember = "CarrierName"
			//Me.carrierListBoxControl.ValueMember = "CarrierID"
			carriersSearchGridControl.DataSource = m_Carriers.GetCarrierIDAndNames();

		}

		private bool SaveChanges()
		{

			//Required to get back to current carrier when entering a new carrier.
			int selectedID = -1;

			if (m_CurrentCarrierID.HasValue == false)
			{
				m_CurrentCarrierID = m_Carriers.GetNewCarrierID();
			}
			else
			{
				selectedID = carriersSearchGridView.FocusedRowHandle;
			}

			try
			{
				string firstName = null;
				string lastName = null;
				if (firstNameTextEdit.EditValue == null || Convert.ToString(firstNameTextEdit.EditValue) == firstNameTextEdit.Properties.NullText)
				{
					firstName = string.Empty;
				}
				else
				{
					firstName = Convert.ToString(firstNameTextEdit.EditValue);
				}
				if (lastNameTextEdit.EditValue == null || Convert.ToString(lastNameTextEdit.EditValue) == lastNameTextEdit.Properties.NullText)
				{
					lastName = string.Empty;
				}
				else
				{
					lastName = Convert.ToString(lastNameTextEdit.EditValue);
				}
				if (m_Carriers.UpdateCarrier(m_CurrentCarrierID.Value, nameTextEdit.Text, firstName, lastName, addressTextEdit.Text, cityTextEdit.Text, stateTextEdit.Text, postalTextEdit.Text, phoneTextEdit.Text, emailTextEdit.Text, faxTextEdit.Text, noteMemoEdit.Text) != true)
				{
					MessageBox.Show("The carrier was not updated succesfully.", "Error Encountered", MessageBoxButtons.OK, MessageBoxIcon.Error);
					return false;
				}
			}
			catch (ApplicationException ex)
			{
				MessageBox.Show(ex.Message, "Error Encountered", MessageBoxButtons.OK, MessageBoxIcon.Error);
				return false;
			}

			BindCarrierSearch();
			carriersSearchGridView.FocusedRowHandle = selectedID;
			return true;

		}

		private void saveSimpleButton_Click(System.Object sender, EventArgs e)
		{

			SaveChanges();

		}

		private void carrierListBoxControl_SelectedValueChanged(System.Object sender, EventArgs e)
		{

			//If Me.carrierListBoxControl.SelectedIndex <> -1 Then
			//    BindCarrierControls(CType(Me.carrierListBoxControl.SelectedValue, Integer))
			//End If

		}

		public void BindCarrierControls(int carrierID)
		{

			SPG.CarriersRow carrier = (SPG.CarriersRow)(m_Carriers.GetCarrierByID(carrierID).Rows[0]);

			m_CurrentCarrierID = carrier.CarrierID;
			nameTextEdit.Text = carrier.CarrierName;
			firstNameTextEdit.Text = carrier.ContactFirstName;
			lastNameTextEdit.Text = carrier.ContactLastName;
			addressTextEdit.Text = carrier.Address;
			cityTextEdit.Text = carrier.City;
			stateTextEdit.Text = carrier.State;
			postalTextEdit.Text = carrier.Postal;
			phoneTextEdit.Text = carrier.Phone;
			emailTextEdit.Text = carrier.Email;
			faxTextEdit.Text = carrier.Fax;
			noteMemoEdit.Text = carrier.Note;

		}

		private void addSimpleButton_Click(System.Object sender, EventArgs e)
		{

			m_CurrentCarrierID = null;
			nameTextEdit.Text = string.Empty;
			firstNameTextEdit.Text = null;
			lastNameTextEdit.Text = null;
			addressTextEdit.Text = string.Empty;
			cityTextEdit.Text = string.Empty;
			stateTextEdit.Text = string.Empty;
			postalTextEdit.Text = string.Empty;
			phoneTextEdit.Text = string.Empty;
			emailTextEdit.Text = string.Empty;
			faxTextEdit.Text = string.Empty;
			noteMemoEdit.Text = string.Empty;

		}

		private void cancelSimpleButton_Click(System.Object sender, EventArgs e)
		{

			//If Me.carrierListBoxControl.SelectedIndex <> -1 Then
			//    BindCarrierControls(CType(Me.carrierListBoxControl.SelectedValue, Integer))
			//End If

		}

		private void carriersSearchGridView_Click(object sender, EventArgs e)
		{

			if (carriersSearchGridView.CalcHitInfo(carriersSearchGridControl.PointToClient(Control.MousePosition)).HitTest == GridHitTest.RowCell)
			{
				BindCarrierControls(Convert.ToInt32(carriersSearchGridView.GetFocusedRowCellValue(carrierIDGridColumn)));
			}

		}

		private void addBarButtonItem_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{

			m_CurrentCarrierID = null;
			nameTextEdit.Text = string.Empty;
			firstNameTextEdit.Text = null;
			lastNameTextEdit.Text = null;
			addressTextEdit.Text = string.Empty;
			cityTextEdit.Text = string.Empty;
			stateTextEdit.Text = string.Empty;
			postalTextEdit.Text = string.Empty;
			phoneTextEdit.Text = string.Empty;
			emailTextEdit.Text = string.Empty;
			faxTextEdit.Text = string.Empty;
			noteMemoEdit.Text = string.Empty;

			carriersSearchGridView.FocusedRowHandle = -1;
			Utilities.MakeFormReadOnly(carriersXtraTabPage, false);
			carriersSearchGridControl.Enabled = false;
			CheckPermissions();

			cancelBarButtonItem.Enabled = true;
			saveBarButtonItem.Enabled = true;
			editBarButtonItem.Enabled = false;
			addBarButtonItem.Enabled = false;

		}

		private void editBarButtonItem_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{

			if (m_CurrentCarrierID == null)
			{
				MessageBox.Show("Please select a carrier to edit.");
				return;
			}

			Utilities.MakeFormReadOnly(carriersXtraTabPage, false);
			carriersSearchGridControl.Enabled = false;
			CheckPermissions();

			cancelBarButtonItem.Enabled = true;
			saveBarButtonItem.Enabled = true;
			editBarButtonItem.Enabled = false;
			addBarButtonItem.Enabled = false;

		}

		private void cancelBarButtonItem_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{

			CancelChanges();

			CheckPermissions();

			Utilities.MakeFormReadOnly(carriersXtraTabPage, true);
			carriersSearchGridControl.Enabled = true;

			cancelBarButtonItem.Enabled = false;
			saveBarButtonItem.Enabled = false;

		}

		private void saveBarButtonItem_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{

			if (SaveChanges())
			{

				CheckPermissions();

				Utilities.MakeFormReadOnly(carriersXtraTabPage, true);
				carriersSearchGridControl.Enabled = true;

				cancelBarButtonItem.Enabled = false;
				saveBarButtonItem.Enabled = false;
			}

		}

		public bool CheckPermissions()
		{

			try
			{
				SPG.UserPermissionsRow permission = (new UserPermissionsBLL()).GetUserPermissionsByUserName(Properties.Settings.Default.UserName, "Carrierss Form")[0];

				switch (permission.PermissionLevel)
				{
					case "FULL":
					case "Full":
						addBarButtonItem.Enabled = true;
						editBarButtonItem.Enabled = true;
						//Me.deleteBarButtonItem.Enabled = True
						break;
					case "ADD":
					case "Add":
						addBarButtonItem.Enabled = true;
						editBarButtonItem.Enabled = true;
						break;
					case "NONE":
					case "None":
						MessageBox.Show("You don't have permission to view this form.", "User Permissions", MessageBoxButtons.OK, MessageBoxIcon.Hand);
						Close();
						return false;
					case "VIEW":
					case "View":
						addBarButtonItem.Enabled = false;
						editBarButtonItem.Enabled = false;
						//Me.deleteBarButtonItem.Enabled = False
						break;
					case "EDIT":
					case "Edit":
						addBarButtonItem.Enabled = false;
						editBarButtonItem.Enabled = true;
						//Me.deleteBarButtonItem.Enabled = False
						break;
				}

				return true;
			}
			catch (InvalidOperationException ex)
			{
				return false;
			}

		}

		private void CancelChanges()
		{

			if (m_CurrentCarrierID != null)
			{
				BindCarrierControls(m_CurrentCarrierID.Value);
			}

		}

	}
}